### ISSUE TITLE
cmd/go: 'vet' and 'test' duplicate flags from GOFLAGS if they are also specified explicitly

### ISSUE THREAD

**Issue by elagergren-spideroak** [golang/go#32471]
### What version of Go are you using (`go version`)?

<pre>
$ go version
1.12.5
</pre>

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
darwin/amd64
</pre></details>

### What did you do?

Set `GOFLAGS=-vet=off` then ran `go test -vet=all`.

### What did you expect to see?

`go test ...` to not error out.

### What did you see instead?

```
$ GOFLAGS=-test.vet=off go test -test.vet all
# _/tmp/bar
./bar_test.go:6:2: Logf format %f has arg X of wrong type string
FAIL	_/tmp/bar [build failed]
```

```
$ GOFLAGS=-vet=off go test -vet all
go test: vet flag may be set only once
run "go help test" or "go help testflag" for more information
```

```
$ GOFLAGS=-test.vet=off go test -vet all
# _/tmp/bar
./bar_test.go:6:2: Logf format %f has arg X of wrong type string
FAIL	_/tmp/bar [build failed]
```

```
$ GOFLAGS=-vet=off go test -test.vet all
go test: vet flag may be set only once
run "go help test" or "go help testflag" for more information
```

**Comment by elagergren-spideroak** [comment]
Similarly, `GOFLAGS="-vet=all -vet=off" go test -v` does not work. Related to https://github.com/golang/go/issues/29053

**Comment by elagergren-spideroak** [comment]
@gopherbot, please add tag WaitingForInfo

**Comment by bcmills** [comment]
What happens for the other flags with the `test.` prefix, such as `-test.v` and `-test.run`?

We should probably at least make `-vet` consistent with those.

**Comment by elagergren-spideroak** [comment]
```
$ GOFLAGS=-test.v go test -test.v
# _/tmp/bar
./bar_test.go:6:2: Logf format %f has arg X of wrong type string
FAIL	_/tmp/bar [build failed]
$ GOFLAGS=-test.v go test -v
# _/tmp/bar
./bar_test.go:6:2: Logf format %f has arg X of wrong type string
FAIL	_/tmp/bar [build failed]
$ GOFLAGS=-v go test -v
go test: v flag may be set only once
run "go help test" or "go help testflag" for more information
$ GOFLAGS=-v go test -test.v
go test: v flag may be set only once
run "go help test" or "go help testflag" for more information
```

**Comment by elagergren-spideroak** [comment]
Run exhibits the same behavior as `-test.v` and `-test.vet`.

**Comment by agnivade** [comment]
The errors seem to be consistent for all flags.

@elagergren-spideroak - It's not clear from the issue description, but did you want `GOFLAGS=-vet=off go test -vet all` the vet=all value to override the vet=off value set in `GOFLAGS` ?

**Comment by amenzhinsky** [comment]
I'm having the same problem with persistent environment variables, even if I set the corresponding env var to an empty string:
```
go env -w GOFLAGS="-trimpath"
GOFLAGS= go test -trimpath # or simply "go test -trimpath"

go test: trimpath flag may be set only once
run "go help test" or "go help testflag" for more information
```

**Comment by amenzhinsky** [comment]
It seems only `go test` and `go vet` have the issue, everything is alright with the others.

**Comment by elagergren-spideroak** [comment]
@agnivade Yeah. From the docs:

```
GOFLAGS
	...
        Flags listed on the command line
	are applied after this list and therefore override it.
```

**Comment by agnivade** [comment]
@amenzhinsky - Rightly noted. This was introduced in [CL 126656](https://go-review.googlesource.com/c/go/+/126656).

Only go/internal/test/testflag.go has `f, value, extraWord := cmdflag.Parse(cmd, usage, testFlagDefn, args, i)`

and go/internal/vet/vetflag.go has `f, value, extraWord := cmdflag.Parse("vet", usage, vetFlagDefn, args, i)` due to which this happens.

The args come from modification with `cmdflag.FindGOFLAGS(vetFlagDefn)`

The other go subcommands have a different code path for flag processing.

Perhaps @rsc @bcmills will know why is this.

**Comment by amenzhinsky** [comment]
I think the easiest fix would be removing this:
 https://github.com/golang/go/blob/99957b6930c76b683dbca1ff4bcdd56e59b1e035/src/cmd/go/internal/cmdflag/flag.go#L124
Moreover all subcommands and tools accept flags overriding (correct me if I'm wrong, checked on that vert quickly).

**Comment by ianlancetaylor** [comment]
@bcmills Is this going to happen for 1.14?

**Comment by bcmills** [comment]
Yes.

**Comment by bcmills** [comment]
A minimal fix for 1.14 is now merged. I plan to follow that up with a cleaner refactoring for 1.15.
